packages/firestore/test/integration/api/get_options.test.ts

```diff
@@ -0,0 +1,612 @@
+              });
+            } else {
+              expect(qrySnap.docChanges.length).to.equal(3);
+              expect(docsData).to.deep.equal({
+                doc2: { key2b: 'value2b' },
+                doc3: { key3b: 'value3b' },
+                doc4: { key4: 'value4' }
+              });
+            }
+          })
+      );
+    });
+  });
+
+  it('get document while online cache only', () => {
+    const initialData = { key: 'value' };
+    return withTestDoc(persistence, docRef => {
+      return docRef
+        .set(initialData)
+        .then(() => docRef.get({ source: 'cache' }))
+        .then(doc => {
+          if (persistence) {
+            expect(doc.exists).to.be.true;
+            expect(doc.metadata.fromCache).to.be.true;
+            expect(doc.metadata.hasPendingWrites).to.be.false;
+            expect(doc.data()).to.deep.equal(initialData);
+          } else {
+            expect.fail();
+          }
+        })
+        .catch(err => {
+          // we expect an error when persistence has been disabled (since the
+          // document won't be in the local cache.)
+          expect(persistence).to.be.false;
+        });
+    });
+  });
+
+  it('get collection while online cache only', () => {
+    const initialDocs = {
+      doc1: { key1: 'value1' },
+      doc2: { key2: 'value2' },
+      doc3: { key3: 'value3' }
+    };
+    return withTestCollection(persistence, initialDocs, colRef => {
+      // force local cache of these
+      return colRef
+        .get()
+        .then(ignored => colRef.get({ source: 'cache' }))
+        .then(qrySnap => {
+          expect(qrySnap.metadata.fromCache).to.be.true;
+          expect(qrySnap.metadata.hasPendingWrites).to.be.false;
+          if (persistence) {
+            expect(qrySnap.docChanges.length).to.equal(3);
+            let docsData = {};
+            qrySnap.forEach(doc => {
+              docsData[doc.id] = doc.data();
+            });
+            expect(docsData).to.deep.equal(initialDocs);
+          } else {
+            expect(qrySnap.docChanges.length).to.equal(0);
+            expect(qrySnap.empty).to.be.true;
+          }
+        });
+    });
+  });
+
+  it('get document while offline cache only', () => {
+    const initialData = { key: 'value' };
+    const newData = { key2: 'value2' };
+    return withTestDoc(persistence, docRef => {
+      return docRef
+        .set(initialData)
+        .then(() => docRef.firestore.disableNetwork())
+        .then(() => {
+          // NB: since we're offline, the returned promise won't complete
+          docRef.set(newData);
+          return docRef.get({ source: 'cache' });
+        })
+        .then(doc => {
+          expect(doc.exists).to.be.true;
+          expect(doc.metadata.fromCache).to.be.true;
+          expect(doc.metadata.hasPendingWrites).to.be.true;
+          expect(doc.data()).to.deep.equal(newData);
+        });
+    });
+  });
+
+  it('get collection while offline cache only', () => {
```

Comment:
FWIW, to make it extra obvious what's being tested, perhaps say "with source=cache" instead of "cache only"